### DISCUSSION TITLE
Array functions for JSON arrays

### DISCUSSION THREAD

**Discussion by NrujalSharma** [sebastienros/jint#1733] [Q&A] [answered]
**Version used**
3.0.0-beta-2059

**Describe the bug**
I've a JSON array whose properties can be accessed without any issue. But trying to access Array functions such as map/filter/reduce it throws the following error.
```
No public methods with the specified arguments were found.
```

**To Reproduce**
## Sample Code
```C#
void Main()
{
	var expression = "participants.find(p => p.id)";
	
	string d = File.ReadAllText(@"<PATH_TO_JSON>\MinimumReproducible.json");
	JObject jObject = JObject.Parse(d);
	
	var eval = new Engine()
	.SetValue("log", new Action<object>(Console.WriteLine))
	.SetValue("participants", jObject["participants"])
	.Execute($"let result = {expression}; log(result)")
	.Evaluate("result");
}

```
## Sample JSON
```json
{
  "participants": [
    {
      "participantType": {
        "id": 1,
        "name": "Client"
      },
      "id": "8d1d2e485aab4c52be5c5926353fb9db"
    },
    {
      "participantType": {
        "id": 4,
        "name": "Vendor"
      },
      "id": "9b68c253d55749ee89a4c993d13e5bc3"
    },
    {
      "participantType": {
        "id": 1,
        "name": "Client"
      },
      "id": "a8f6ab053b9944edb5a5982c897c70e1"
    }
  ]
}
```

**Expected behavior**
Array methods such as map, reduce, filter should be available for use on JSON arrays

**Additional context**
I tried this using both inbuilt JSON as well as NewtonSoft JSON library. The array methods cannot be accessed with either.

**Comment by lahma** [comment, marked as answer]
You need to configure your engine to allow specific types to get array prototype, see https://github.com/sebastienros/jint/blob/main/Jint.Tests/Runtime/InteropTests.SystemTextJson.cs for an example, you don't need to set `TypeResolver` with Newtonsoft.JSON, it's an STJ workaround.

**Reply by NrujalSharma** [reply]
@lahma Thank you for your response

Please consider the following code:
```C#
void Main()
{
	var expression = "participants.find(p => p.id)";
	
	string d = File.ReadAllText(@"C:\Users\nrujal.sharma\Downloads\MinimumReproducible.json");
	
	var eval = new Engine()
	.SetValue("log", new Action<object>(Console.WriteLine))
	.Execute(@$"
		const data = {d};
		const participants = data.participants;
	")
	.Execute($"let result = {expression}; log(result)")
	.Evaluate("result");
}
```

Here, I took all the JSON parsing away from the C# code, supplied the JSON directly to the engine for execution.
Is it okay to do this or will it lead to some other issues?

**Reply by lahma** [reply]
That should be quite fine, it'll be native JavaScript object then.
